# English Premier League 2019-2020

The following Notebook is a statistical analysis for the 2019-2020 English Premier League.

The __Player__ dataset was taken from Kaggle.com

Link to the dataset: https://www.kaggle.com/idoyo92/epl-stats-20192020 (https://www.kaggle.com/idoyo92/epl-stats-20192020?select=players_1920_fin.csv)

The __Match__ dataset was taken from Football-Data.co.uk
Link to the dataset: http://www.football-data.co.uk/englandm.php (http://www.football-data.co.uk/mmz4281/1920/E0.csv)

---

In [57]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
# pd.set_option('display.width', 500)
# pd.set_option('display.max_columns', 100)

## Dataset 1: Match Data

First we will organise the data for football matches

---

In [65]:
# Read the data from the dataset into a dataframe
incomplete_matches_df = pd.read_csv('epl-dataset/epl2020.csv')

In [59]:
incomplete_matches_df.head()

Unnamed: 0 h_a        xG       xGA      npxG     npxGA  deep  deep_allowed  scored  missed    xpts result                 date  wins  draws  loses  pts     npxGD       teamId   ppda_cal  allowed_ppda  matchtime  tot_points  round  tot_goal  tot_con Referee.x  HS.x  HST.x  HF.x  HC.x  HY.x  HR.x  AS.x  AST.x  AF.x  AC.x  AY.x  AR.x  B365H.x  B365D.x  B365A.x  HtrgPerc  AtrgPerc matchDay
0           1   h  2.234560  0.842407  2.234560  0.842407    11             5       4       1  2.3863      w  2019-08-09 20:00:00     1      0      0    3  1.392153    Liverpool   5.942857     21.454545         20           3      1         4        1  M Oliver    15      7     9    11     0     0    12      5     9     2     2     0     1.14     10.0    19.00  0.466667  0.416667      Fri
1           2   a  0.842407  2.234560  0.842407  2.234560     5            11       1       4  0.4405      l  2019-08-09 20:00:00     0      0      1    0 -1.392153      Norwich  21.454545      5.942857         20           0      1         1        4  M Oliver    15      7     9    11     0     0    12      5     9     2     2     0     1.14     10.0    19.00  0.466667  0.416667      Fri
2           3   a  3.183770  1.200300  2.422640  1.200300     9             1       5       0  2.6312      w  2019-08-10 12:30:00     1      0      0    3  1.222340     Man City   6.935484     16.400000         12           3      1         5        0    M Dean     5      3     6     1     2     0    14      9    13     1     2     0    12.00      6.5     1.22  0.600000  0.642857      Sat
3           4   h  1.200300  3.183770  1.200300  2.422640     1             9       0       5  0.2522      l  2019-08-10 12:30:00     0      0      1    0 -1.222340     West Ham  16.400000      6.935484         12           0      1         0        5    M Dean     5      3     6     1     2     0    14      9    13     1     2     0    12.00      6.5     1.22  0.600000  0.642857      Sat
4           5   h  1.340990  1.598640  1.340990  1.598640     4             6       1       1  1.0172      d  2019-08-10 15:00:00     0      1      0    1 -0.257650  Bournemouth  13.600000      9.133333         15           1      1         1        1  K Friend    13      3    10     3     2     0     8      3    19     4     1     0     1.95      3.6     3.60  0.230769  0.375000      Sat

In the following result, we can see there are 576 rows

In [60]:
incomplete_matches_df.shape

(576, 45)

In [61]:
incomplete_matches_df.shape[0] / 2 # number of matches

288.0

In [62]:
# Display the last 5 rows
incomplete_matches_df.iloc[::-1].head()

Unnamed: 0 h_a        xG       xGA      npxG     npxGA  deep  deep_allowed  scored  missed    xpts result                 date  wins  draws  loses  pts     npxGD       teamId   ppda_cal  allowed_ppda  matchtime  tot_points  round  tot_goal  tot_con Referee.x  HS.x  HST.x  HF.x  HC.x  HY.x  HR.x  AS.x  AST.x  AF.x  AC.x  AY.x  AR.x  B365H.x  B365D.x  B365A.x  HtrgPerc  AtrgPerc matchDay
575         576   h  2.412390  0.672448  1.651220  0.672448     8             2       4       0  2.5950      w  2020-03-09 20:00:00     1      0      0    3  0.978772    Leicester   5.379310     13.217391         20          53     29        58       28  M Oliver    15      7    15     9     2     0     4      1    12     0     1     0     1.36     5.25      8.0  0.466667  0.250000      Mon
574         575   a  0.672448  2.412390  0.672448  1.651220     2             8       0       4  0.2673      l  2020-03-09 20:00:00     0      0      1    0 -0.978772  Aston Villa  13.217391      5.379310         20          25     28        34       56  M Oliver    15      7    15     9     2     0     4      1    12     0     1     0     1.36     5.25      8.0  0.466667  0.250000      Mon
573         574   h  1.738900  0.585469  1.738900  0.585469     3             2       2       0  2.4269      w  2020-03-08 16:30:00     1      0      0    3  1.153431      Man Utd  14.709677      6.041667         16          45     29        44       30    M Dean    12      6    11     2     2     0     7      2     9    11     4     0     5.00     4.33      1.6  0.500000  0.285714      Sun
572         573   a  0.585469  1.738900  0.585469  1.738900     2             3       0       2  0.3539      l  2020-03-08 16:30:00     0      0      1    0 -1.153431     Man City   6.041667     14.709677         16          57     28        68       31    M Dean    12      6    11     2     2     0     7      2     9    11     4     0     5.00     4.33      1.6  0.500000  0.285714      Sun
571         572   a  0.445922  2.028830  0.445922  2.028830     3             9       0       4  0.2901      l  2020-03-08 14:00:00     0      0      1    0 -1.582908      Everton  19.086957     12.714286         14          37     29        37       46  K Friend    17     11     8     6     1     0     3      1    10     1     2     0     1.80     3.90      4.2  0.647059  0.333333      Sun

In [63]:
# Get the date for the last match
incomplete_matches_df.iloc[-1].date

'2020-03-09 20:00:00'

This has led to discovery that the dataset is incomplete. A full season should contain a total of __380__ matches.
The shape of the data above indicates there are only 288 (576 / 2) matches contained in the dataset. 
Upon manual inspection of the data, we can see that the date for the last match in the dataset occurs on the 9th of March 2020. This coincides with when the Premier League was postponed due to Covid. The League resumed several weeks later, but the dataset does not include the final matches.

---

### Alternative Football Match Dataset

http://www.football-data.co.uk/mmz4281/1920/E0.csv

Link to the column names: http://www.football-data.co.uk/notes.txt

### Exploring The Data

In [458]:
matches_df = pd.read_csv('epl-dataset/matches.csv')

In [459]:
matches_df.shape

(380, 106)

In [460]:
matches_df.shape[0]

380

__380__ rows = 380 matches - The expected amount

---

In [461]:
matches_df.head()

Div        Date   Time        HomeTeam          AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR   Referee  HS  AS  HST  AST  HF  AF  HC  AC  HY  AY  HR  AR  B365H  B365D  B365A    BWH   BWD    BWA    IWH   IWD    IWA    PSH   PSD    PSA    WHH  WHD    WHA    VCH  VCD    VCA   MaxH   MaxD   MaxA   AvgH  AvgD   AvgA  B365>2.5  B365<2.5  ...   AHh  B365AHH  B365AHA  PAHH  PAHA  MaxAHH  MaxAHA  AvgAHH  AvgAHA  B365CH  B365CD  B365CA   BWCH  BWCD   BWCA   IWCH  IWCD   IWCA   PSCH   PSCD   PSCA   WHCH  \
0  E0  09/08/2019  20:00       Liverpool           Norwich     4     1   H     4     0   H  M Oliver  15  12    7    5   9   9  11   2   0   2   0   0   1.14  10.00  19.00   1.14  8.25  18.50   1.15  8.00  18.00   1.15  9.59  18.05   1.12  8.5  21.00   1.14  9.5  23.00   1.16  10.00  23.00   1.14  8.75  19.83      1.40      3.00  ... -2.25     1.96     1.94  1.97  1.95    1.97    2.00    1.94    1.94    1.14    9.50   21.00   1.14   9.0  20.00   1.15  8.00  18.00   1.14  10.43  19.63   1.11   
1  E0  10/08/2019  12:30        West Ham          Man City     0     5   A     0     1   A    M Dean   5  14    3    9   6  13   1   1   2   2   0   0  12.00   6.50   1.22  11.50  5.75   1.26  11.00  6.10   1.25  11.68  6.53   1.26  13.00  6.0   1.24  12.00  6.5   1.25  13.00   6.75   1.29  11.84  6.28   1.25      1.44      2.75  ...  1.75     2.00     1.90  2.02  1.90    2.02    1.92    1.99    1.89   12.00    7.00    1.25  11.00   6.0   1.26  11.00  6.10   1.25  11.11   6.68   1.27  11.00   
2  E0  10/08/2019  15:00     Bournemouth  Sheffield United     1     1   D     0     0   D  K Friend  13   8    3    3  10  19   3   4   2   1   0   0   1.95   3.60   3.60   1.95  3.60   3.90   1.97  3.55   3.80   2.04  3.57   3.90   2.00  3.5   3.80   2.00  3.6   4.00   2.06   3.65   4.00   2.01  3.53   3.83      1.90      1.90  ... -0.50     2.01     1.89  2.04  1.88    2.04    1.91    2.00    1.88    1.95    3.70    4.20   1.95   3.6   3.90   1.97  3.55   3.85   1.98   3.67   4.06   1.95   
3  E0  10/08/2019  15:00         Burnley       Southampton     3     0   H     0     0   D   G Scott  10  11    4    3   6  12   2   7   0   0   0   0   2.62   3.20   2.75   2.65  3.20   2.75   2.65  3.20   2.75   2.71  3.31   2.81   2.70  3.2   2.75   2.70  3.3   2.80   2.80   3.33   2.85   2.68  3.22   2.78      2.10      1.72  ...  0.00     1.92     1.98  1.93  2.00    1.94    2.00    1.91    1.98    2.70    3.25    2.90   2.65   3.1   2.85   2.60  3.20   2.85   2.71   3.19   2.90   2.62   
4  E0  10/08/2019  15:00  Crystal Palace           Everton     0     0   D     0     0   D    J Moss   6  10    2    3  16  14   6   2   2   1   0   1   3.00   3.25   2.37   3.20  3.20   2.35   3.10  3.20   2.40   3.21  3.37   2.39   3.10  3.3   2.35   3.20  3.3   2.45   3.21   3.40   2.52   3.13  3.27   2.40      2.20      1.66  ...  0.25     1.85     2.05  1.88  2.05    1.88    2.09    1.84    2.04    3.40    3.50    2.25   3.30   3.3   2.25   3.40  3.30   2.20   3.37   3.45   2.27   3.30   

   WHCD   WHCA   VCCH  VCCD   VCCA  MaxCH  MaxCD  MaxCA  AvgCH  AvgCD  AvgCA  B365C>2.5  B365C<2.5  PC>2.5  PC<2.5  MaxC>2.5  MaxC<2.5  AvgC>2.5  AvgC<2.5  AHCh  B365CAHH  B365CAHA  PCAHH  PCAHA  MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  
0   9.5  21.00   1.14  9.50  23.00   1.16  10.50  23.00   1.14   9.52  19.18        1.3       3.50    1.34    3.44      1.36      3.76      1.32      3.43 -2.25      1.91      1.99   1.94   1.98     1.99     2.07     1.90     1.99  
1   6.5   1.24  12.00  6.50   1.25  13.00   7.00   1.29  11.14   6.46   1.26        1.4       3.00    1.43    3.03      1.50      3.22      1.41      2.91  1.75      1.95      1.95   1.96   1.97     2.07     1.98     1.97     1.92  
2   3.6   3.90   2.00  3.60   4.00   2.03   3.70   4.20   1.98   3.58   3.96        1.9       1.90    1.94    1.97      1.97      1.98      1.91      1.92 -0.50      1.95      1.95   1.98   1.95     2.00     1.96     1.96     1.92  
3   3.2   2.80   2.70  3.25   2.90   2.72   3.26   2.95   2.65   3.18   2.8

In [462]:
matches_df.iloc[::-1].head()

Div        Date   Time     HomeTeam          AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR     Referee  HS  AS  HST  AST  HF  AF  HC  AC  HY  AY  HR  AR  B365H  B365D  B365A   BWH    BWD    BWA   IWH    IWD    IWA   PSH    PSD    PSA   WHH   WHD    WHA   VCH   VCD    VCA  MaxH   MaxD   MaxA  AvgH   AvgD   AvgA  B365>2.5  B365<2.5  ...   AHh  B365AHH  B365AHA  PAHH  PAHA  MaxAHH  MaxAHA  AvgAHH  AvgAHA  B365CH  B365CD  B365CA  BWCH  BWCD   BWCA  IWCH   IWCD   IWCA  PSCH   PSCD   PSCA  WHCH  \
379  E0  26/07/2020  16:00     West Ham       Aston Villa     1     1   D     0     0   D    M Oliver  10  13    1    4  16  13   0   7   2   1   0   0   3.00   3.60   2.30  3.00   3.50   2.30  3.05   3.40   2.30  3.12   3.62   2.31  3.00   3.6   2.30  3.10   3.6   2.25  3.20   3.82   2.36  3.05   3.57   2.28      1.66      2.20  ...  0.25     1.91     1.99  1.92  2.01    1.94    2.02    1.89    1.97    2.55     3.5    2.62  2.55   3.6   2.65  2.60   3.35   2.65  2.66   3.53   2.69  2.62   
378  E0  26/07/2020  16:00  Southampton  Sheffield United     3     1   H     0     1   A    P Bankes  13   5    4    3   9  16   9   1   0   1   0   0   2.15   3.50   3.30  2.20   3.50   3.20  2.20   3.30   3.35  2.18   3.54   3.46  2.15   3.4   3.40  2.15   3.4   3.40  2.25   3.60   3.56  2.18   3.47   3.34      1.90      1.90  ... -0.25     1.89     2.01  1.89  2.04    1.92    2.04    1.88    1.99    2.00     3.5    3.80  2.00   3.5   3.75  2.00   3.40   3.75  2.00   3.68   3.93  2.00   
377  E0  26/07/2020  16:00    Newcastle         Liverpool     1     3   A     1     1   D    A Taylor   3  14    2    6  11   5   2   4   1   0   0   0   7.50   5.25   1.36  7.50   4.75   1.42  7.50   4.80   1.40  7.79   5.07   1.41  8.00   5.0   1.38  8.00   4.8   1.40  8.20   5.40   1.45  7.60   4.99   1.40      1.53      2.50  ...  1.50     1.75     2.05  1.83  2.10    1.87    2.14    1.79    2.08    5.50     4.6    1.53  4.75   4.5   1.62  5.25   4.20   1.60  5.42   4.62   1.58  6.00   
376  E0  26/07/2020  16:00     Man City           Norwich     5     0   H     2     0   H    C Pawson  31   5   10    4   7   4   9   0   1   1   0   0   1.08  11.00  26.00  1.05  13.50  41.00  1.07  12.00  30.00  1.08  13.10  30.22  1.07  15.0  21.00  1.05  13.0  34.00  1.10  15.25  41.00  1.07  12.80  30.42      1.22      4.33  ... -3.00     1.95     1.95  1.93  1.97    2.00    2.00    1.95    1.92    1.03    19.0   41.00  1.02  19.5  51.00  1.07  14.00  28.00  1.04  18.50  41.00  1.02   
375  E0  26/07/2020  16:00    Leicester        Man United     0     2   A     0     0   D  M Atkinson  14   7    3    3  12  11   3   3   1   4   1   0   3.25   3.50   2.20  3.20   3.50   2.20  3.15   3.25   2.30  3.25   3.51   2.28  3.20   3.4   2.25  3.20   3.4   2.25  3.36   4.54   2.35  3.17   3.49   2.24      1.90      1.90  ...  0.25     1.94     1.42  1.94  1.98    1.98    2.01    1.91    1.92    3.60     3.5    2.05  3.70   3.6   2.00  3.65   3.40   2.05  3.76   3.65   2.05  3.80   

     WHCD   WHCA  VCCH  VCCD  VCCA  MaxCH  MaxCD  MaxCA  AvgCH  AvgCD  AvgCA  B365C>2.5  B365C<2.5  PC>2.5  PC<2.5  MaxC>2.5  MaxC<2.5  AvgC>2.5  AvgC<2.5  AHCh  B365CAHH  B365CAHA  PCAHH  PCAHA  MaxCAHH  MaxCAHA  AvgCAHH  AvgCAHA  
379   3.6   2.55  2.63   3.5   2.7   2.75   3.76   2.75   2.62   3.50   2.65       1.80       2.00    1.86    2.05      1.93      2.13      1.80      2.03  0.00      1.93      1.97   1.95   1.98     1.99     2.00     1.93     1.95  
378   3.6   3.75  1.95   3.6   4.0   2.05   3.85   4.07   1.99   3.61   3.81       1.80       2.00    1.83    2.09      1.90      2.11      1.82      2.01 -0.50      1.97      1.93   2.00   1.93     2.03     1.96     1.98     1.89  
377   4.4   1.52  5.75   4.1   1.6   6.45   4.80   1.62   5.57   4.48   1.56       1.57       2.37    1.62    2.44      1.62      2.51      1.57      2.40  1.00      1.94      1.96   1.95   1.97     2.03     2.00     1.95     1.92  
376  21.0  34.00  1.03  23.0  34.0   1.07  26.00  58.00   1.04  19.42  40.86       1.12       6

In [463]:
matches_df.iloc[-1].Date

'26/07/2020'

This is the correct end date for the 2019-2020 Premier League. Last match: Sunday, 26 July 2020

---

### Removing Unneeded Columns

The dataset contains a lot of columns in relation to betting odds. This data is not needed for the analysis being conducted here, so this data can be removed.

In [464]:
# matches_df.drop([
#     'Div',
#     'B365H',
#     'B365D',
#     'B365A',
#     'BSH',
#     'BSD',
#     'BSA',
#     'BWH',
#     'BWD',
#     'BWA',
#     'GBH',
#     'GBD',
#     'GBA',
#     'IWH',
#     'IWD',
#     'IWA',
#     'LBH',  
# ], axis=1, inplace=True, errors='ignore')

Rather than removing columns, I thought it would be best to select the columns needed, as there are many more columns that I do not need. There is a lot more work to list the columns to remove.

In [465]:
matches_df = matches_df[[
    'Date',
    'Time', 
    'HomeTeam', 
    'AwayTeam', 
    'FTHG', 
    'FTAG', 
    'FTR', 
    'HTHG',
    'HTAG',
    'HTR',
    'Referee',
    'HS',
    'AS',
    'HST',
    'AST',
    'HF',
    'AF',
    'HC',
    'AC',
    'HY',
    'AY',
    'HR',
    'AR'
]]

In [466]:
matches_df.head()

Date   Time        HomeTeam          AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR   Referee  HS  AS  HST  AST  HF  AF  HC  AC  HY  AY  HR  AR
0  09/08/2019  20:00       Liverpool           Norwich     4     1   H     4     0   H  M Oliver  15  12    7    5   9   9  11   2   0   2   0   0
1  10/08/2019  12:30        West Ham          Man City     0     5   A     0     1   A    M Dean   5  14    3    9   6  13   1   1   2   2   0   0
2  10/08/2019  15:00     Bournemouth  Sheffield United     1     1   D     0     0   D  K Friend  13   8    3    3  10  19   3   4   2   1   0   0
3  10/08/2019  15:00         Burnley       Southampton     3     0   H     0     0   D   G Scott  10  11    4    3   6  12   2   7   0   0   0   0
4  10/08/2019  15:00  Crystal Palace           Everton     0     0   D     0     0   D    J Moss   6  10    2    3  16  14   6   2   2   1   0   1

---

### Renaming Columns

In [467]:
# http://www.football-data.co.uk/notes.txt

matches_df = matches_df.rename(columns={
    'Date': 'date',
    'Time': 'time',
    'HomeTeam': 'home_team',
    'AwayTeam': 'away_team',
    'Referee': 'referee',
    'FTHG': 'full_time_home_goals',
    'FTAG': 'full_time_away_goals',
    'FTR': 'full_time_result',
    'HTHG': 'half_time_home_goals',
    'HTAG': 'half_time_away_goals',
    'HTR': 'half_time_result',
    'HS': 'home_shots',
    'AS': 'away_shots',
    'HST': 'home_shots_on_target',
    'AST': 'away_shots_on_target',
    'HF': 'home_fouls',
    'AF': 'away_fouls',
    'HC': 'home_corners',
    'AC': 'away_corners',
    'HY': 'home_yellow_cards',
    'AY': 'away_yellow_cards',
    'HR': 'home_red_cards',
    'AR': 'away_red_cards'
})

In [468]:
matches_df.head()

date   time       home_team         away_team  full_time_home_goals  full_time_away_goals full_time_result  half_time_home_goals  half_time_away_goals half_time_result   referee  home_shots  away_shots  home_shots_on_target  away_shots_on_target  home_fouls  away_fouls  home_corners  away_corners  home_yellow_cards  away_yellow_cards  home_red_cards  away_red_cards
0  09/08/2019  20:00       Liverpool           Norwich                     4                     1                H                     4                     0                H  M Oliver          15          12                     7                     5           9           9            11             2                  0                  2               0               0
1  10/08/2019  12:30        West Ham          Man City                     0                     5                A                     0                     1                A    M Dean           5          14                     3                     9           6          13             1             1                  2                  2               0               0
2  10/08/2019  15:00     Bournemouth  Sheffield United                     1                     1                D                     0                     0                D  K Friend          13           8                     3                     3          10          19             3             4                  2                  1               0               0
3  10/08/2019  15:00         Burnley       Southampton                     3                     0                H                     0                     0                D   G Scott          10          11                     4                     3           6          12             2             7                  0                  0               0               0
4  10/08/2019  15:00  Crystal Palace           Everton                     0                     0                D                     0                     0                D    J Moss           6          10                     2                     3          16          14             6             2                  2                  1               0               1

---

### Check for missing values (null values)

Looking at the overall column stats provided by Kaggle, they provide some __total__ stats on each column within the dataset which allows us to identify any columns that may contain invalid data. And vice verse, we are able to identify columns that contain only valid data, eg. 1 and 0 or Y and N. This enables us to figure out which columns will need to be tested. Unfortunatel with the new dataset from Football-Data.co.uk, there are no summary totals for the columns. Only the raw CSV is provided. We will need to perform some checks on the columns we need to ensure the data is complete and valid.

When we explore the Player data from Kaggle, these column summary totals will enable us to skip the need to check for null values within certain columns that appear to be complete.
We will still need to confirm there are no null values for the string value columns, such as the referee names, team names, day of the week, etc.

In [469]:
print(np.sum([matches_df.date.isnull()]))
print(np.sum([matches_df.time.isnull()]))
print(np.sum([matches_df.home_team.isnull()]))
print(np.sum([matches_df.away_team.isnull()]))
print(np.sum([matches_df.full_time_home_goals.isnull()]))
print(np.sum([matches_df.full_time_away_goals.isnull()]))
print(np.sum([matches_df.full_time_result.isnull()]))
print(np.sum([matches_df.half_time_home_goals.isnull()]))
print(np.sum([matches_df.half_time_away_goals.isnull()]))
print(np.sum([matches_df.half_time_result.isnull()]))
print(np.sum([matches_df.referee.isnull()]))
print(np.sum([matches_df.home_shots.isnull()]))
print(np.sum([matches_df.away_shots.isnull()]))
print(np.sum([matches_df.home_shots_on_target.isnull()]))
print(np.sum([matches_df.away_shots_on_target.isnull()]))
print(np.sum([matches_df.home_fouls.isnull()]))
print(np.sum([matches_df.away_fouls.isnull()]))
print(np.sum([matches_df.home_corners.isnull()]))
print(np.sum([matches_df.away_corners.isnull()]))
print(np.sum([matches_df.home_yellow_cards.isnull()]))
print(np.sum([matches_df.away_yellow_cards.isnull()]))
print(np.sum([matches_df.home_red_cards.isnull()]))
print(np.sum([matches_df.away_red_cards.isnull()]))

0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0


---

### Ensure Data Types Are Correct

In [470]:
matches_df.dtypes

date                    object
time                    object
home_team               object
away_team               object
full_time_home_goals     int64
full_time_away_goals     int64
full_time_result        object
half_time_home_goals     int64
half_time_away_goals     int64
half_time_result        object
referee                 object
home_shots               int64
away_shots               int64
home_shots_on_target     int64
away_shots_on_target     int64
home_fouls               int64
away_fouls               int64
home_corners             int64
away_corners             int64
home_yellow_cards        int64
away_yellow_cards        int64
home_red_cards           int64
away_red_cards           int64
dtype: object

Most of the column datatypes in the dataset are correct. There are mostly integer values, eg. __full_time_home_goals__.

The __home_team__, __away_team__ and __referee__ columns are all strings (object). The __full_time_result__ and __half_time_result__ are single character strings (eg. A or H).

We need to change the type for the __date__ field in particular to a datetime64 type. 

We will also need to format the date to ensure the days and months are consistent. Without formatting, it will automatically assign the day as the month and vice versa.

In [471]:
# matches_df.Date = matches_df.Date.astype('datetime64[ns]')

# Format date to ensure consistency with days and months
matches_df.date = pd.to_datetime(matches_df.date.astype(str), format='%d/%m/%Y') 

In [472]:
# Format time
matches_df.time = pd.to_datetime(matches_df.time.astype('datetime64[ns]')) 
matches_df.time = matches_df.time.dt.time

In [473]:
matches_df.dtypes

date                    datetime64[ns]
time                            object
home_team                       object
away_team                       object
full_time_home_goals             int64
full_time_away_goals             int64
full_time_result                object
half_time_home_goals             int64
half_time_away_goals             int64
half_time_result                object
referee                         object
home_shots                       int64
away_shots                       int64
home_shots_on_target             int64
away_shots_on_target             int64
home_fouls                       int64
away_fouls                       int64
home_corners                     int64
away_corners                     int64
home_yellow_cards                int64
away_yellow_cards                int64
home_red_cards                   int64
away_red_cards                   int64
dtype: object

In [474]:
matches_df.head()

date      time       home_team         away_team  full_time_home_goals  full_time_away_goals full_time_result  half_time_home_goals  half_time_away_goals half_time_result   referee  home_shots  away_shots  home_shots_on_target  away_shots_on_target  home_fouls  away_fouls  home_corners  away_corners  home_yellow_cards  away_yellow_cards  home_red_cards  away_red_cards
0 2019-08-09  20:00:00       Liverpool           Norwich                     4                     1                H                     4                     0                H  M Oliver          15          12                     7                     5           9           9            11             2                  0                  2               0               0
1 2019-08-10  12:30:00        West Ham          Man City                     0                     5                A                     0                     1                A    M Dean           5          14                     3                     9           6          13             1             1                  2                  2               0               0
2 2019-08-10  15:00:00     Bournemouth  Sheffield United                     1                     1                D                     0                     0                D  K Friend          13           8                     3                     3          10          19             3             4                  2                  1               0               0
3 2019-08-10  15:00:00         Burnley       Southampton                     3                     0                H                     0                     0                D   G Scott          10          11                     4                     3           6          12             2             7                  0                  0               0               0
4 2019-08-10  15:00:00  Crystal Palace           Everton                     0                     0                D                     0                     0                D    J Moss           6          10                     2                     3          16          14             6             2                  2                  1               0               1

---

### Move Date and Time Columns

There is no specific reason to move the date and time columns other than for cleanliness. The next step is to break out the dates and times into separate columns (day, month, year, etc.) so it will be cleaner to keep them at the end of the dataframe.

In [475]:
# Move Date and Time to the end of the dataframe

date = matches_df.date
time = matches_df.time

matches_df.drop(['date', 'time'], axis=1, inplace=True, errors='ignore')

matches_df['date'] = date
matches_df['time'] = time

In [476]:
matches_df.head()

home_team         away_team  full_time_home_goals  full_time_away_goals full_time_result  half_time_home_goals  half_time_away_goals half_time_result   referee  home_shots  away_shots  home_shots_on_target  away_shots_on_target  home_fouls  away_fouls  home_corners  away_corners  home_yellow_cards  away_yellow_cards  home_red_cards  away_red_cards       date      time
0       Liverpool           Norwich                     4                     1                H                     4                     0                H  M Oliver          15          12                     7                     5           9           9            11             2                  0                  2               0               0 2019-08-09  20:00:00
1        West Ham          Man City                     0                     5                A                     0                     1                A    M Dean           5          14                     3                     9           6          13             1             1                  2                  2               0               0 2019-08-10  12:30:00
2     Bournemouth  Sheffield United                     1                     1                D                     0                     0                D  K Friend          13           8                     3                     3          10          19             3             4                  2                  1               0               0 2019-08-10  15:00:00
3         Burnley       Southampton                     3                     0                H                     0                     0                D   G Scott          10          11                     4                     3           6          12             2             7                  0                  0               0               0 2019-08-10  15:00:00
4  Crystal Palace           Everton                     0                     0                D                     0                     0                D    J Moss           6          10                     2                     3          16          14             6             2                  2                  1               0               1 2019-08-10  15:00:00

---

### Separate Date Values

Create new columns for Year and Month

In [477]:
matches_df['year'] = matches_df.date.dt.year
matches_df['month'] = matches_df.date.dt.month
matches_df['day'] = matches_df.date.dt.day

In [478]:
matches_df.head()

home_team         away_team  full_time_home_goals  full_time_away_goals full_time_result  half_time_home_goals  half_time_away_goals half_time_result   referee  home_shots  away_shots  home_shots_on_target  away_shots_on_target  home_fouls  away_fouls  home_corners  away_corners  home_yellow_cards  away_yellow_cards  home_red_cards  away_red_cards       date      time  year  month  day
0       Liverpool           Norwich                     4                     1                H                     4                     0                H  M Oliver          15          12                     7                     5           9           9            11             2                  0                  2               0               0 2019-08-09  20:00:00  2019      8    9
1        West Ham          Man City                     0                     5                A                     0                     1                A    M Dean           5          14                     3                     9           6          13             1             1                  2                  2               0               0 2019-08-10  12:30:00  2019      8   10
2     Bournemouth  Sheffield United                     1                     1                D                     0                     0                D  K Friend          13           8                     3                     3          10          19             3             4                  2                  1               0               0 2019-08-10  15:00:00  2019      8   10
3         Burnley       Southampton                     3                     0                H                     0                     0                D   G Scott          10          11                     4                     3           6          12             2             7                  0                  0               0               0 2019-08-10  15:00:00  2019      8   10
4  Crystal Palace           Everton                     0                     0                D                     0                     0                D    J Moss           6          10                     2                     3          16          14             6             2                  2                  1               0               1 2019-08-10  15:00:00  2019      8   10

---
---
---

## Dataset 2: Player Data